### PR TITLE
Fix: China Overlord, Emperor model, texture, Gattling issues

### PR DESCRIPTION
**Pull request by xezon**
This change fixes all jank on Overlord and Emperor models.

Fixes:
* All muzzle flashes are viewable from all sides
* Gattling Cannon muzzle flashes do show now
* Gattling Cannon barrels rotate with proper speeds
* Gattling Cannon stops rotating on cease fire
* Chassis holes on bottom of Overlord closed
* Frankenstein Emperor models replaced with new model
* ~~Removed odd wheels from Overlord chassis textures~~

Does not fix:
* Installing vehicle addon on REALLYDAMAGED vehicle does not spawn damaged addon variant

All W3D Ini changes have been copied as is from NProject. They have been tested and found working. The Emperor Tank has some additional setup merged from Crazy Mod setup due new Emperor model.

New Emperor model uses same textures as original Overlord and Emperor do. The original Emperor is a mash up of Overlord parts poorly put together, with stretched textures, poorly intersecting meshes, treads texture on the top of the chassis, and visible bones on the damaged model variant.

The Emperor model was created by Jordan, Crazy Mod.
https://www.moddb.com/mods/cc-generals-crazy-mod/images/china-emperor-tank-re-make

### TODO

- [ ] Remove all formatting changes
- [ ] Recreate model fixes

# Overlord

## Original

shot_20220918_162112_11

## Patched, NProject

shot_20220918_161359_7

# Emperor

## Original

shot_20220918_161738_1

shot_20220918_161840_2

shot_20220918_161905_3

shot_20220918_161912_4

shot_20220918_163154_3

## Patched, CrazyMod

shot_20220918_083131_1

shot_20220918_161225_3

### CODE REVIEW

**Comment by commy2** [comment]
Why all those unnecessary config changes? This will make it harder to compare against 1.04.

**Comment by xezon** [comment]
As stated above, I copied the setups from NProject. I have no good understanding of what is happening in them. If it can be written differently, we can do that, but it works as is so that is good enough for me.

**Comment by commy2** [comment]
All changes in W3DOverlordTankDraw are idiosyncratic formatting and thus unnecessary.

The changes for Gatling upgrade FX are good, but should go into their own pull request.

This Emperor model should be optional if anything. The excessive team color looks pretty ugly to me at first glance.

**Comment by xezon** [comment]
I can't do this change better than this. Emperor cannot be Optional because of Bones mismatch. The original Emperor cannot be preferable, because it is broken beyond rescue. If you can fix House Color, then you can do that.

**Comment by ExiLe-ZH** [comment]
> This Emperor model should be optional if anything. The excessive team color looks pretty ugly to me at first glance.

+1

**Comment by xezon** [comment]
This change will hit main branch unless someone can make it better. I understand you don't like changes, but we cannot keep the original jank. It does not look good.

**Comment by ExiLe-ZH** [comment]
99.9% of ppl never noticed.

new ugly model is instantly noticed by everyone.

**Comment by commy2** [comment]
At least separate out the Gatling turret changes from the Emperor changes. And definitely don't mass replace formatting. I will look into this when I have more time and separate them.

**Comment by xezon** [comment]
I tried separating Emperor and Overlord change, but it did not work somehow. Using new Overlord model(s) breaks the Gattling Cannon on legacy Emperor Tank somehow. This is why I combined the changes, as this gives Emperor its own new Gattling Cannon which works in return. I did separate the Nuke Overlord however, as that was no problem.

**Comment by xezon** [comment]
As for the Emperor Model, it is huge visual upgrade to original one. This is obvious in many details. That you do not like big house color ring on turret is unfortunate. I don't mind it, looks ok to me. When zooming out and looking at new Emperor Tank in regular camera height, it is also not that large really.

Top New Emperor, Bottom New Overlord with Speaker.

shot_20220918_195658_1

File: Patch104pZH/GameFilesEdited/Data/INI/Object/NukeGeneral.ini [hunk 1]
```diff
@@ -16712,41 +16712,36 @@ Object Nuke_ChinaTankOverlord
...
+    DefaultConditionState
+      Model             = NVOvrlrd
+      Animation         = NVOvrlrd.NVOvrlrd
+      AnimationMode     = LOOP
+      Turret            = Turret01
+      WeaponFireFXBone  = PRIMARY Muzzle
+      WeaponRecoilBone  = PRIMARY Barrel
+      WeaponMuzzleFlash = PRIMARY MuzzleFX
+      WeaponLaunchBone  = PRIMARY Muzzle
+      HideSubObject     = NukeUpgrade ; Hide white hinges
```
Comment (commy2) [inline Patch104pZH/GameFilesEdited/Data/INI/Object/NukeGeneral.ini:16728]: ?

Comment (xezon) [reply Patch104pZH/GameFilesEdited/Data/INI/Object/NukeGeneral.ini:16728]: That is not supposed to be there. I will remove it.

File: Patch104pZH/GameFilesEdited/Data/INI/Object/TankGeneral.ini [hunk 2]
```diff
@@ -3333,29 +3443,21 @@ Object Tank_ChinaTankOverlordPropagandaTower
   ;UpgradeCameo4 = NONE
   UpgradeCameo5 = Upgrade_ChinaOverlordPropagandaTower
 
-  Draw                    = W3DDependencyModelDraw ModuleTag_01
-    OkToChangeModelColor  = Yes
-    AttachToBoneInContainer = FIREPOINT01
+  Draw = W3DDependencyModelDraw ModuleTag_01
+    OkToChangeModelColor = Yes
+    AttachToBoneInContainer = FirePoint01
```
Comment (commy2) [inline Patch104pZH/GameFilesEdited/Data/INI/Object/TankGeneral.ini:3448]: This object is unused in 1.04. What does this do here?

File: Patch104pZH/GameFilesEdited/Data/INI/Object/TankGeneral.ini [hunk 3]
```diff
@@ -3438,33 +3540,17 @@ Object Tank_ChinaTankOverlordBattleBunker
   ;UpgradeCameo4 = NONE
   UpgradeCameo5 = Upgrade_ChinaOverlordBattleBunker
 
-  Draw                    = W3DDependencyModelDraw ModuleTag_01
-    OkToChangeModelColor  = Yes
-    AttachToBoneInContainer = FIREPOINT01
+  Draw = W3DDependencyModelDraw ModuleTag_01
```
Comment (commy2) [inline Patch104pZH/GameFilesEdited/Data/INI/Object/TankGeneral.ini:3543]: Emperor Battle Bunker? ???

File: Patch104pZH/GameFilesEdited/Data/INI/Object/ChinaCINEUnit.ini [hunk 4]
```diff
@@ -1238,41 +1238,35 @@ Object CINE_ChinaTankOverlord
...
 
-    TrackMarks           = EXTnkTrack.tga
-    OkToChangeModelColor = Yes
-    TreadAnimationRate = 2.0;  amount of tread texture to move per second
+    TreadAnimationRate      = 2.0  ; amount of tread texture to move per second
     TreadDriveSpeedFraction = 0.3  ; fraction of locomotor speed below which treads stop moving.
     TreadPivotSpeedFraction = 0.6  ; fraction of locomotor speed below which we allow pivoting.
+
+    TrackMarks = EXTnkTrack.tga
+
```
Comment (commy2) [inline Patch104pZH/GameFilesEdited/Data/INI/Object/ChinaCINEUnit.ini:1269]: This whole block is the same content, but shuffled around. I think we should avoid large formatting changes that have no affect at this stage to keep code compareable. This entire hunk can be dropped.

File: Patch104pZH/GameFilesEdited/Data/INI/Object/ChinaVehicle.ini [hunk 5]
```diff
@@ -210,41 +210,35 @@ Object ChinaTankOverlord
...
-    TrackMarks           = EXTnkTrack.tga
-    OkToChangeModelColor = Yes
-    TreadAnimationRate = 2.0;  amount of tread texture to move per second
+    TreadAnimationRate      = 2.0  ; amount of tread texture to move per second
     TreadDriveSpeedFraction = 0.3  ; fraction of locomotor speed below which treads stop moving.
     TreadPivotSpeedFraction = 0.6  ; fraction of locomotor speed below which we allow pivoting.
+
+    TrackMarks = EXTnkTrack.tga
+
   End
```
Comment (commy2) [inline Patch104pZH/GameFilesEdited/Data/INI/Object/ChinaVehicle.ini:242]: Same as above.

File: Patch104pZH/GameFilesEdited/Data/INI/Object/ChinaVehicle.ini [hunk 6]
```diff
@@ -458,74 +452,170 @@ Object ChinaTankOverlordGattlingCannon
   ;UpgradeCameo4 = NONE
   UpgradeCameo5 = Upgrade_ChinaOverlordGattlingCannon
 
+  ; Patch104p, NProject @bugfix properly implement Gattling Cannon to show muzzle flashes and smoke particles
```
Comment (commy2) [inline Patch104pZH/GameFilesEdited/Data/INI/Object/ChinaVehicle.ini:455]: Comment should be neutral. The previous implementation is just as "proper".
```suggestion
  ; Patch104p, NProject @bugfix Implement Gattling Cannon muzzle flashes and smoke particles
```

File: Patch104pZH/GameFilesEdited/Data/INI/Object/NukeGeneral.ini [hunk 7]
```diff
@@ -16712,41 +16712,36 @@ Object Nuke_ChinaTankOverlord
...
     End
 
-    TrackMarks           = EXTnkTrack.tga
-    OkToChangeModelColor = Yes
-    TreadAnimationRate = 2.0;  amount of tread texture to move per second
+    TreadAnimationRate      = 2.0  ; amount of tread texture to move per second
     TreadDriveSpeedFraction = 0.3  ; fraction of locomotor speed below which treads stop moving.
     TreadPivotSpeedFraction = 0.6  ; fraction of locomotor speed below which we allow pivoting.
+
+    TrackMarks = EXTnkTrack.tga
```
Comment (commy2) [inline Patch104pZH/GameFilesEdited/Data/INI/Object/NukeGeneral.ini:16745]: Same as 2 above. Except that strange HideSubObject thing.

File: Patch104pZH/GameFilesEdited/Data/INI/Object/NukeGeneral.ini [hunk 8]
```diff
@@ -16934,74 +16929,170 @@ Object Nuke_ChinaTankOverlordGattlingCannon
   ;UpgradeCameo4 = NONE
   UpgradeCameo5 = Upgrade_ChinaOverlordGattlingCannon
 
+  ; Patch104p, NProject @bugfix properly implement Gattling Cannon to show muzzle flashes and smoke particles
```
Comment (commy2) [inline Patch104pZH/GameFilesEdited/Data/INI/Object/NukeGeneral.ini:16934]: above

File: Patch104pZH/GameFilesEdited/Data/INI/Object/NukeGeneral.ini [hunk 9]
```diff
@@ -16934,74 +16929,170 @@ Object Nuke_ChinaTankOverlordGattlingCannon
...
+      Animation           = NVOvrlrd_G.NVOvrlrd_G
+      AnimationMode       = MANUAL
+      Turret              = Turret01
+      TurretPitch         = TurretEL
       WeaponFireFXBone    = PRIMARY Muzzle
       WeaponMuzzleFlash   = PRIMARY MuzzleFX
+    End
+    ConditionState        = USING_WEAPON_B
       WeaponFireFXBone    = SECONDARY Muzzle
       WeaponMuzzleFlash   = SECONDARY MuzzleFX
```
Comment (commy2) [inline Patch104pZH/GameFilesEdited/Data/INI/Object/NukeGeneral.ini:16952]: I _think_ I know why this is used. Would have to check if it is really necessary to show muzzle flash and light ("fx") when firing up. I only ever encountered issues with WeaponRecoilBone, not these two.

File: Patch104pZH/GameFilesEdited/Data/INI/Object/NukeGeneral.ini [hunk 10]
```diff
@@ -17085,29 +17176,21 @@ Object Nuke_ChinaTankOverlordPropagandaTower
...
-      ParticleSysBone = Spark01 LiveWireSparks
-    End
-    ; night
-    ConditionState = NIGHT
-      Model         = NVOvrlrd_S
-      Animation     = NVOvrlrd_S.NVOvrlrd_S
+      Animation     = NVOvrlrd_SD.NVOvrlrd_SD
       AnimationMode = LOOP
     End
+
```
Comment (commy2) [inline Patch104pZH/GameFilesEdited/Data/INI/Object/NukeGeneral.ini:17195]: This hunk doesn't do anything.

File: Patch104pZH/GameFilesEdited/Data/INI/Object/NukeGeneral.ini [hunk 11]
```diff
@@ -17185,33 +17268,18 @@ Object Nuke_ChinaTankOverlordBattleBunker
...
-    ConditionState = NIGHT REALLYDAMAGED
-      Model         = NVOvrlrd_BD
-      ParticleSysBone = Smoke01 SmolderingSmoke
-      ParticleSysBone = Smoke02 SmolderingSmoke
-      ParticleSysBone = Flame01 SmolderingFire
-      ParticleSysBone = Flame01 SmolderingFlameCore
-      ParticleSysBone = Spark01 LiveWireSparks
+      Model = NVOvrlrd_BD
     End
+
```
Comment (commy2) [inline Patch104pZH/GameFilesEdited/Data/INI/Object/NukeGeneral.ini:17284]: Where do all the smoke particles go? Are they not spawned for OverlordContain cargo? If so, this is code cleanup and imo out of scope for this, as it is already a large PR.

File: Patch104pZH/GameFilesEdited/Data/INI/Object/TankGeneral.ini [hunk 12]
```diff
@@ -3177,74 +3190,171 @@ Object Tank_ChinaTankOverlordGattlingCannon
   ;UpgradeCameo4 = NONE
   UpgradeCameo5 = Upgrade_ChinaOverlordGattlingCannon
 
+  ; Patch104p, NProject @bugfix properly implement Gattling Cannon to show muzzle flashes and smoke particles
```
Comment (commy2) [inline Patch104pZH/GameFilesEdited/Data/INI/Object/TankGeneral.ini:3193]: Same as above.

**Comment by MTKing4** [comment]
Agree, separate bug/texture fixes from new model, put new model optional

**Comment by xezon** [comment]
Texture changes have been isolated from this change.

**Comment by xezon** [comment]
We can likely separate some or all Gattling Cannon fixes from this change.

**Comment by xezon** [comment]
This change has been superseeded for the most part. Overlord, Emperor model fixes shall be carried out in the future. Emperor needs remodel or reskin.

**Pull request closed without merging**